### PR TITLE
Get required image info for UpgradeLog from cluster repo

### PR DESCRIPTION
**Pull request by starbops**
**IMPORTANT: Please do not create a Pull Request without creating an issue first.**

**Problem:**

We hardcoded the UpgradeLog-related image repos and tags in the controller code base.

**Solution:**

Determine the image repos and tags by querying the corresponding charts from the cluster repo container at runtime.

**Related Issue:**

#3595 

**Test plan:**

1. Build a new ISO image containing this PR and provision a single node cluster with it
2. Upgrade the cluster with the same ISO image
3. Check the `UpgradeLog` CR to see if `InfraReady` condition is `True`
4. After the `DownloadReady` becomes `True`, try to download the log archive

### CODE REVIEW

File: pkg/controller/master/upgradelog/controller.go [hunk 1]
```diff
@@ -146,7 +145,12 @@ func (h *handler) OnUpgradeLogChange(_ string, upgradeLog *harvesterv1.UpgradeLo
 			return nil, err
 		}
 		// The creation of the Logging resource will indirectly bring up fluent-bit DaemonSet and fluentd StatefulSet
-		if _, err := h.loggingClient.Create(prepareLogging(upgradeLog)); err != nil && !apierrors.IsAlreadyExists(err) {
+		url := fmt.Sprintf("http://%s.%s/%s", util.HarvesterClusterRepoSvcName, util.CattleSystemNamespaceName, rancherLoggingChartValuesPath)
```
Comment (w13915984028) [inline pkg/controller/master/upgradelog/controller.go]: The default images in `values.yaml` may be overwritten by pre-defined values from `managedchart` /  `addons`, and if the users modify the values in `managedchart` / `addon`, that will be the final values to be used.

Seems we need to also pick values from `managedchart` /  `addons` to overwrite the default ones as following.

```
images:
  config_reloader:
    repository: rancher/mirrored-jimmidyson-configmap-reload
    tag: v0.4.0
  fluentbit:
    repository: rancher/mirrored-fluent-fluent-bit
    tag: 1.9.3
  nodeagent_fluentbit:
    os: "windows"
    repository: rancher/fluent-bit
    tag: 1.9.3
  fluentbit_debug:
    repository: rancher/mirrored-fluent-fluent-bit
    tag: 1.9.3-debug
  fluentd:
    repository: rancher/mirrored-banzaicloud-fluentd
    tag: v1.14.6-alpine-5
```

File: pkg/controller/master/upgradelog/common.go [hunk 2]
```diff
@@ -96,7 +96,13 @@ func prepareOperator(upgradeLog *harvesterv1.UpgradeLog) *mgmtv3.ManagedChart {
 	}
 }
 
-func prepareLogging(upgradeLog *harvesterv1.UpgradeLog) *loggingv1.Logging {
+func getImageInfo(imageName, key string, images map[string]interface{}) string {
+	image := images[imageName].(map[string]interface{})
```
Comment (w13915984028) [inline pkg/controller/master/upgradelog/common.go]: if `image` is nil, will `image[key]` break?

**Review by w13915984028 — COMMENTED** [review]
please double check, thanks.

**Comment by chrisho** [comment]
@starbops Could you rebase the latest master?

**Review by chrisho — APPROVED** [review]
LGTM and the case seem to work fine, please rebase the master and fix the conflicts, thanks.
```yaml
apiVersion: harvesterhci.io/v1beta1
kind: UpgradeLog
metadata:
  annotations:
    harvesterhci.io/clusterFlow: ClusterFlowReady
    harvesterhci.io/clusterOutput: ClusterOutputReady
    harvesterhci.io/fluentBit: FluentBitReady
    harvesterhci.io/fluentd: FluentdReady
    harvesterhci.io/upgradeLogState: Stopped
  creationTimestamp: "2023-06-30T11:24:51Z"
  finalizers:
  - wrangler.cattle.io/harvester-upgradelog-controller
  generation: 11
  labels:
    harvesterhci.io/upgrade: hvst-upgrade-nk84n
  name: hvst-upgrade-nk84n-upgradelog
  namespace: harvester-system
  ownerReferences:
  - apiVersion: harvesterhci.io/v1beta1
    kind: Upgrade
    name: hvst-upgrade-nk84n
    uid: 0f8d6ec7-f837-455c-9dee-42f39e85c0c0
  resourceVersion: "28766"
  uid: 0644f74f-1b6b-478a-9419-13efc333bfe2
spec:
  upgradeName: hvst-upgrade-nk84n
status:
  conditions:
  - lastUpdateTime: "2023-06-30T11:25:18Z"
    status: "True"
    type: Started
  - lastUpdateTime: "2023-06-30T11:24:58Z"
    status: "True"
    type: LoggingOperatorReady
  - lastUpdateTime: "2023-06-30T11:25:15Z"
    status: "True"
    type: InfraReady
  - lastUpdateTime: "2023-06-30T11:33:03Z"
    status: "True"
    type: Stopped
  - lastUpdateTime: "2023-06-30T11:25:24Z"
    status: "True"
    type: DownloadReady
```
image
image

File: pkg/controller/master/upgradelog/common.go (hunk 2 above)
Comment (Vicente-Cheng) [reply pkg/controller/master/upgradelog/common.go]: Could we check the map size first?
Also, would we need to check if the key cannot be found on the images?

File: pkg/controller/master/upgradelog/controller.go (hunk 1 above)
Comment (starbops) [reply pkg/controller/master/upgradelog/controller.go]: Thanks for the heads-up. Now the implementation becomes fetching image info from `apps.catalog.cattle.io/v1` and consolidates the default values with the user-defined values.

Comment (bk201) [reply pkg/controller/master/upgradelog/controller.go]: cc @chrisho, now it's get from apps, which contains the final value of a chart.

File: pkg/controller/master/upgradelog/common.go [hunk 3]
```diff
@@ -96,7 +125,13 @@ func prepareOperator(upgradeLog *harvesterv1.UpgradeLog) *mgmtv3.ManagedChart {
 	}
 }
 
-func prepareLogging(upgradeLog *harvesterv1.UpgradeLog) *loggingv1.Logging {
+// func getImageInfo(imageName, key string, images map[string]interface{}) string {
```
Comment (bk201) [inline pkg/controller/master/upgradelog/common.go]: Please remove unused code.

File: pkg/controller/master/upgradelog/common.go [hunk 4]
```diff
@@ -32,6 +32,35 @@ func upgradeLogReference(upgradeLog *harvesterv1.UpgradeLog) metav1.OwnerReferen
 	}
 }
 
+func getLoggingImages(candidateImages map[string]image, imageList map[string]interface{}) error {
```
Comment (bk201) [inline pkg/controller/master/upgradelog/common.go]: @starbops Can you check if this function can be used here? https://github.com/harvester/harvester/blob/8feea706be8bc84b8c30b2f6b5e45c3861345a65/pkg/controller/master/setting/support_bundle_image.go#L23

Comment (starbops) [reply pkg/controller/master/upgradelog/common.go]: Updated, thanks for the suggestion!

**Review by chrisho — APPROVED** [review]
The latest commit's code looks more readable and works normally through the verified.
Here is a suggestion, add a progression for Upgradelog on the upgrade tooltips, it will appear Upgradelog's status, ensure that it lets users know what's next and even if it encounters an error and show it up. 
Thanks.

**Review by Vicente-Cheng — APPROVED** [review]
LGTM, thanks!

**Review by bk201 — APPROVED** [review]
LGTM. Test with
- logging add-on disabled.
- logging add-on enabled.
- logging add-on enabled, and override values with
```
  valuesContent: |
    images:
      fluentbit:
        repository: fluent/fluent-bit
        tag: 1.9.6
```
The `upgradelog-infra` logging CR can get correct image.

**Comment by bk201** [comment]
@mergifyio backport v1.2

**Pull request merged**